src/sync.cpp

```diff
@@ -2190,33 +2201,172 @@ void UnifiedSync::changedConfigState(bool notifyApp)
+                << result;
+    }
+    else
+    {
+        LOG_verbose << "Internal config database flushed to disk.";
+    }
+
+    return result;
+}
+
+error Syncs::syncConfigDBLoad()
+{
+    LOG_verbose << "Attempting to load internal sync configs from disk.";
+
+    auto result = API_EAGAIN;
+
+    // Can we get our hands on the internal sync config database?
+    if (auto* db = syncConfigDB())
+    {
+        // Make sure we have a suitable IO context.
+        if (auto* ioContext = syncConfigIOContext())
+        {
+            // Try and read the database from disk.
+            result = db->read(*ioContext);
+        }
+
+        if (result == API_OK || result == API_ENOENT)
+        {
+            LOG_verbose << "Loaded "
+                        << db->configs().size()
+                        << " internal sync config(s) from disk.";
+
+            return API_OK;
+        }
+    }
+
+    LOG_err << "Couldn't load internal sync configs from disk: "
+            << result;
+
+    return result;
+}
+
+JSONSyncConfigIOContext* Syncs::syncConfigIOContext()
+{
+    // Has a suitable IO context already been created?
+    if (mSyncConfigIOContext)
+    {
+        // Yep, return a reference to it.
+        return mSyncConfigIOContext.get();
+    }
+
+    using KeyStr  = Base64Str<SymmCipher::KEYLENGTH * 2>;
+    using NameStr = Base64Str<SymmCipher::KEYLENGTH>;
+
+    if (User* selfUser = mClient.finduser(mClient.me))
+    {
+        auto n = selfUser->getattr(ATTR_JSON_SYNC_CONFIG_NAME);
+        auto k = selfUser->getattr(ATTR_JSON_SYNC_CONFIG_KEY);
+
+        if (n && k &&
+            k->size() == KeyStr::STRLEN &&
+            n->size() == NameStr::STRLEN)
+        {
+            // Create the IO context.
+            mSyncConfigIOContext.reset(
+              new JSONSyncConfigIOContext(mClient.key,
+                                         *mClient.fsaccess,
+                                         *k,
+                                         *n,
+                                         mClient.rng));
+        }
+    }
+    // Return a reference to the new IO context.
+    return mSyncConfigIOContext.get();
+}
+
 void Syncs::clear()
 {
+    syncConfigDBFlush();
+
+    mSyncConfigDB.reset();
+    mSyncConfigIOContext.reset();
     mSyncVec.clear();
-    resetSyncConfigDb();
     isEmpty = true;
 }
 
-void Syncs::resetSyncConfigDb()
+error Syncs::truncate()
 {
-    mSyncConfigDb.reset();
-    if (mClient.dbaccess && mClient.loggedin() == FULLACCOUNT)
+    if (!mSyncConfigDB)
     {
-        mSyncConfigDb.reset(new SyncConfigBag{ *mClient.dbaccess, *mClient.fsaccess, mClient.rng, mClient.uid });
+        return API_OK;
     }
+
+    mSyncConfigDB->clear();
+
+    return syncConfigDBFlush();
+}
+
+void Syncs::resetSyncConfigDb()
+{
+    mSyncConfigDB.reset();
+    static_cast<void>(syncConfigDB());
 }
 
 auto Syncs::appendNewSync(const SyncConfig& c, MegaClient& mc) -> UnifiedSync*
```

Comment:
This function probably used to return a type that was a member of Syncs, so it would have needed a qualified name to declare the function in that way:  `Syncs::ReturnType Syncs::function(etc)`, whereas in C++11 we can name the return type inside the function where the members of Syncs are already in scope, saving the Syncs:: qualifier.    That syntax is also the one adopted for specifying lambda return types so it's good to be familiar with it.

Reply:
I'm familiar with the syntax for lambdas, it just puzzled me here :-) Thanks for the note